### PR TITLE
Update parseTemplates to filter by tagName

### PR DESCRIPTION
**Pull request by ventuno**
As part of the work in https://github.com/ember-template-lint/ember-template-lint/pull/2483 we want to improve `parseTemplates` so that it filters matched templates based on a list of possible import paths and identifiers.

For example, `parseTemplates` currently identifies `indentedThing` as a template in the following code:
```
export default class Foo extends Service {
  indentedThing = stripIndent`
    Hahaha, this is just a plain string. Definitely not a template.
  `
}
```

We want improve `parseTemplates` so that given the following template and a list of identifiers `[{ importPath: 'ember-cli-handlebars', importIdentifier: 'hbs' }]`:
```
import { hbs } from 'ember-cli-handlebars`;

export default class Foo extends Service {
  indentedThing = stripIndent`
    Hahaha, this is just a plain string. Definitely not a template.
  `
  myTemplate = hbs`
    Hahaha, this is just a plain string. Definitely not a template.
  `
}
```
only `myTemplate` is returned. This makes it easier for `ember-template-lint` to lint templates defined in a script file.

For further details see:
* https://github.com/ember-template-lint/ember-template-lint/pull/2483#discussion_r858969319;
* [Discord thread](https://discord.com/channels/480462759797063690/967432082630541354/969026993310031962).

### CODE REVIEW

File: src/parse-templates.ts [hunk 1]
```diff
@@ -66,14 +77,27 @@ function isEscaped(template: string, _offset: number | undefined) {
 export function parseTemplates(
   template: string,
   relativePath: string,
-  templateTag?: string
+  options?: string | ParseTemplatesOptions
```
Comment (rwjblue) [inline src/parse-templates.ts:80]: I think we mostly expect that this is private API (@dfreeman / @chriskrycho can hopefully confirm), I’d rather just change to taking options and not support the string fallback. 

What do you think?

**Review by rwjblue — COMMENTED** [review]
Looks good overall! One minor question / possible tweak?

File: src/parse-templates.ts (hunk 1 above)
Comment (ventuno) [reply src/parse-templates.ts:80]: I much rather prefer to just change the API if possible, but of course only if we're not going to break anyone else. If someone can confirm, I can make that change quick.

**Comment by NullVoxPopuli** [comment]
I'm excited to use this to add gjs/gts, hbs\`\` support to https://github.com/html-next/svg-icon-optimizer :tada:

File: src/parse-templates.ts (hunk 1 above)
Comment (dfreeman) [reply src/parse-templates.ts:80]: The only external consumer I _know_ of for this is the alpha-quality [`ember-template-imports` Glint environment](https://github.com/typed-ember/glint-environment-ember-template-imports/tree/initial-implementation) (which @ef4 is [working on upstreaming into Glint proper](https://github.com/typed-ember/glint/pull/301)).

I don't think it would be hard for us to absorb this there (or rather, just put out a release that requires a version of `ember-template-imports` that includes this change), but it raises the point that the public interface of this package is pretty nebulous, and there are clearly real-world use cases for using `parseTemplates` elsewhere.

Should we consider exporting this from the package root and making it more officially public?

Comment (ventuno) [reply src/parse-templates.ts]: @rwjblue, @chriskrycho: based on Dan's comment above I went ahead and:
1) bumped the major version in `package.json`;
2) removed support for the string fallback.

Let me know if this works for you.

Comment (chriskrycho) [reply src/parse-templates.ts]: @dfreeman Yeah, I think you’re right: having a public API for various tools will be important. I’m in support of making it public.

@ventuno that seems good to me. 👍🏼 Thanks!

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ember-template-imports",
-  "version": "2.0.1",
+  "version": "3.0.1",
```
Comment (rwjblue) [inline package.json:3]: In general, bumping versions here should be done upon release. Can you revert this change?

**Review by rwjblue — CHANGES_REQUESTED** [review]
One inline tweak, and:

* Figure out what was causing the CI failures
* Can you add a re-export of parseTemplates to the main entry point (src/index.ts)
* Can you add a set of test cases that emulates what we intend to be doing over in ember-template-lint (including the import config that we’ll be using)?
*

**Comment by rwjblue** [comment]
> Can you add a re-export of parseTemplates to the main entry point (src/index.ts)

Specifically, I think we need to:

* create `src/index.ts` with a re-export of `parseTemplates`
* move the top level `index.js` to `ember-addon-main.js`
* update `package.json` to have an `ember-addon.main` that points at `ember-addon-main.js`
* add a `main` entry point that points at `lib/index.js`
* add a `types` entry that points `lib/index.d.ts`

**Comment by ventuno** [comment]
> One inline tweak, and:
> 
>     * Figure out what was causing the CI failures

This is done.

> 
>     * Can you add a re-export of parseTemplates to the main entry point (src/index.ts)
> 

Will work on this next.

>     * Can you add a set of test cases that emulates what we intend to be doing over in ember-template-lint (including the import config that we’ll be using)?

Added tests for each example in https://github.com/emberjs/babel-plugin-ember-template-compilation/blob/9338b0156f3b350a4f2ea201c436aae7741b12b2/src/plugin.ts

**Comment by ventuno** [comment]
> Specifically, I think we need to:

Thanks for the detailed instructions. This is done as well!

File: package.json (hunk 2 above)
Comment (ventuno) [reply package.json]: Updated. Thanks!

**Review by chriskrycho — CHANGES_REQUESTED** [review]

File: package.json [hunk 3]
```diff
@@ -103,12 +104,15 @@
...
   "ember-addon": {
+    "main": "ember-addon-main.js",
     "configPath": "tests/dummy/config",
     "before": [
       "ember-cli-htmlbars",
       "ember-cli-babel"
     ]
   },
+  "main": "lib/index.js",
+  "types": "lib/index.d.ts",
```
[the remark this answers is not in the document]
Comment (rwjblue) [reply package.json:115]: > we can work around that (by adding export declare const hbs: ... to src/index.ts or something like that)

Ya, lets do this

**Review by rwjblue — COMMENTED** [review]
Overall LGTM, I think the last things are:

* fixup linting failure (other tests seem to pass)
* add documentation for the two interfaces that @chriskrycho mentioned
* add `export const hbs =` to the new `src/index.ts` (so that existing users don't start getting issues)

**Comment by dfreeman** [comment]
> add export const hbs = to the new src/index.ts (so that existing users don't start getting issues)

I think we may want to hold on this for now. Leaving it out won't disrupt the status quo (anyone who's been using this with TS either has their own declaration or is using Glint's), and we likely want to be careful to declare it in a way that's still augmentable by Glint in the future to avoid causing clashes for anyone currently using the two together. I'm happy to tackle that in a separate PR.

**Comment by ventuno** [comment]
> Overall LGTM, I think the last things are:
> 
>     * fixup linting failure (other tests seem to pass)

Done. I had to update `eslintrc.js` to match the `index.js` -> `ember-addon-main.js` rename as well.

>     * add documentation for the two interfaces that @chriskrycho mentioned

Done. Hopefully this is a bit clearer now. Happy to iterate.

>     * add `export const hbs =` to the new `src/index.ts` (so that existing users don't start getting issues)

Skipped this based on [@dfreeman's comment](https://github.com/ember-template-imports/ember-template-imports/pull/37#issuecomment-1122548666)

File: src/parse-templates.ts [hunk 4]
```diff
@@ -16,6 +17,16 @@ export interface TemplateLiteralMatch {
...
 
+export interface StaticImportConfig {
+  importPath: string;
+  importIdentifier: string;
+}
+
+export interface ParseTemplatesOptions {
+  templateTag?: string;
+  templateLiteral?: StaticImportConfig[];
+}
```
[the remark this answers is not in the document]
Comment (ventuno) [reply src/parse-templates.ts]: Done. I hope this is a bit clearer now. More than happy to iterate.

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export { parseTemplates } from './parse-templates';
```
Comment (rwjblue) [inline src/index.ts:1]: I think we are still missing https://github.com/ember-template-imports/ember-template-imports/pull/37/files#r869321310 here (publishing as is, will plausibly give folks that are already using typescript an issue).

Comment (rwjblue) [reply src/index.ts:1]: nvm, I see https://github.com/ember-template-imports/ember-template-imports/pull/37#issuecomment-1122548666 now

**Review by rwjblue — APPROVED** [review]

File: src/parse-templates.ts [hunk 6]
```diff
@@ -17,11 +17,23 @@ export interface TemplateLiteralMatch {
...
+/**
+ * Represents a static import of a template literal.
+ *
+ * @param importPath the path to the package from which we want to import the template literal (e.g.: 'ember-cli-htmlbars')
+ * @param importIdentifier the name of the template literal (e.g.: 'hbs') or 'default' if this package exports a default function
+ */
 export interface StaticImportConfig {
   importPath: string;
   importIdentifier: string;
 }
```
Comment (chriskrycho) [inline src/parse-templates.ts:29]: Tweak: this is a place to annotate the properties rather than to use `@param`: as is it won't show up in editors etc. in a useful way.

```ts
/**
 * Represents a static import of a template literal.
 */
export interface StaticImportConfig {
  /**
   * The path to the package from which we want to import the template literal
   * (e.g.: 'ember-cli-htmlbars')
   */
  importPath: string;
  /**
   * The name of the template literal (e.g.: 'hbs') or 'default' if this package
   * exports a default function
   */
  importIdentifier: string;
}
```

File: src/parse-templates.ts [hunk 7]
```diff
@@ -17,11 +17,23 @@ export interface TemplateLiteralMatch {
...
 
+/**
+ * The input options to instruct parseTemplates on how to parse the input.
+ *
+ * @param templateTag tag if parsing template tags is enabled
+ * @param templateLiteral an array of StaticImportConfig to indicate which static imports are expected in this template
+ */
```
Comment (chriskrycho) [inline src/parse-templates.ts:36]: Similarly:

```ts
/**
 * The input options to instruct parseTemplates on how to parse the input.
 *
 * @param templateTag 
 * @param templateLiteral 
 */
export interface ParseTemplatesOptions {
  /** Tag to use, if parsing template tags is enabled. */
  templateTag?: string;
  /** Which static imports are expected in this template. */
  templateLiteral?: StaticImportConfig[];
}
```

File: src/parse-templates.ts [hunk 8]
```diff
@@ -71,7 +83,7 @@ function isEscaped(template: string, _offset: number | undefined) {
  *
  * @param template The template to parse
  * @param relativePath Relative file path for the template (for errors)
- * @param templateTag Optional template tag if parsing template tags is enabled
+ * @param options Optional {ParseTemplateOptions}
```
Comment (chriskrycho) [inline src/parse-templates.ts:86]: Since this is in TS, it doesn't need the JSDoc style attribute, but it's worth documenting anyway:

```suggestion
 * @param options optional configuration options for how to parse templates
```

**Review by chriskrycho — CHANGES_REQUESTED** [review]
Needs a couple small tweaks on the docs and then 👍🏼

File: src/parse-templates.ts (hunk 7 above)
Comment (chriskrycho) [reply src/parse-templates.ts:36]: Non-blocking: whenever I see two optional things like this, it makes me think it should *probably* actually be a union type where *neither* is optional in their own type:

```ts
type TemplateTagOption = {
  /** Tag to use, if parsing template tags is enabled. */
  templateTag: string;
}

type TemplateLiteralOption = {
  /** Which static imports are expected in this template. */
  templateLiteral: StaticImportConfig[];
}

type ParseTemplatesOptions = TemplateTagOption | TemplateLiteralOption;
```

The idea is: it's completely useless to pass `{}` here: it doesn't *do* anything and doesn't even inform what you mean on the caller side. So, instead, using a(n optional) union type says the actual intent, which is "if you pass something it has to be one or the other of these"—at least, if I’ve understood it correctly and it's not legitimate to pass *both*.

Comment (ventuno) [reply src/parse-templates.ts:36]: @chriskrycho my expectations is that we can actually send _both_, as I'm not sure if we can easily determine whether to expect the file will use template tags or template literal. One possible improvement would be to make `templateLiteral` _required_, as it can just be an empty array. Thoughts?

File: src/parse-templates.ts (hunk 6 above)
Comment (ventuno) [reply src/parse-templates.ts]: Done. Thanks!

File: src/parse-templates.ts (hunk 7 above)
Comment (chriskrycho) [reply src/parse-templates.ts]: I’m fine with it if they actually are both optional! It’s just not usually the case. Not blocking though, to be sure.

**Review by chriskrycho — APPROVED** [review]

**Comment by chriskrycho** [comment]
Given this is failing on master in the same way, it's clearly dependency drift nonsense, so we should not block this PR on it. Merging! We'll get the rest sorted separately.
